src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1072 @@
+                    match rx.try_recv() {
+                        Ok(Msg::Close) => break,
+                        Ok(Msg::Disable) => {
+                            enabled = false;
+                            checker.reset_statistics();
+                        }
+                        Ok(Msg::Enable) => {
+                            enabled = true;
+                        }
+                        Err(_) => {}
+                    }
+
+                    if !enabled {
+                        // do nothing, just consume the flow info channel
+                        let _ = flow_info_receiver.recv();
+                        continue;
+                    }
+
+                    match flow_info_receiver.recv_deadline(deadline) {
+                        Ok(FlowInfo::L0(cf, l0_bytes)) => {
+                            if let Some(throttle_cf) = checker.throttle_cf.as_ref() {
+                                if throttle_cf == &cf {
+                                    spare_ticks = 0;
+                                }
+                            }
+                            checker.check_l0_files(cf, l0_bytes)
+                        }
+                        Ok(FlowInfo::L0Intra(cf, diff_bytes)) => {
+                            if let Some(throttle_cf) = checker.throttle_cf.as_ref() {
+                                if throttle_cf == &cf {
+                                    spare_ticks = 0;
+                                }
+                            }
+                            if diff_bytes > 0 {
+                                // Intra L0 merges some deletion records, so regard it as a L0 compaction.
+                                checker.check_l0_files(cf, diff_bytes)
+                            }
+                        }
+                        Ok(FlowInfo::Flush(cf, flush_bytes)) => {
+                            if let Some(throttle_cf) = checker.throttle_cf.as_ref() {
+                                if throttle_cf == &cf {
+                                    spare_ticks = 0;
+                                }
+                            }
+                            checker.adjust_memtables(&cf);
+                            checker.check_flush_flow(cf, flush_bytes)
+                        }
+                        Ok(FlowInfo::Compaction(cf)) => {
+                            if let Some(throttle_cf) = checker.throttle_cf.as_ref() {
+                                if throttle_cf == &cf {
+                                    spare_ticks = 0;
+                                }
+                            }
+                            checker.adjust_pending_compaction_bytes(cf);
+                        }
+                        Err(RecvTimeoutError::Timeout) => {
+                            spare_ticks += 1;
+                            if spare_ticks == SPARE_TICKS_THRESHOLD {
+                                // there is no flush/compaction happens, we should speed up if throttled
```

Comment:
Could you explain again why this happens? Compaction is blocked somehow?

Reply:
the write is stopped, so there is nearly no flush and compaction happens.

Reply:
Why is that? I haven't seen compaction paused because foreground stops. And even if that's the case we shouldn't allow foreground writes when stall condition is still there.

Reply:
For auto-tuned limiter though, should merge this one to avoid similar issue: https://github.com/tikv/rocksdb/pull/234